client/src/main/java/org/daisy/validator/Util.java

```diff
@@ -95,6 +96,19 @@ public static String getRelativeFilename(String currentfile, String link) throws
         return linkPath.substring(parentPath.length());
     }
 
+    public static boolean checkIfContainsDoctype(InputStream inputStream) throws Exception {
+        BufferedReader br = new BufferedReader(new InputStreamReader(inputStream));
+        String line;
+        String fileContent = "";
+        while ((line = br.readLine()) != null) {
+            fileContent += line;
+            if (fileContent.contains("<!DOCTYPE html>")) {
+                return true;
+            }
+        }
```

Comment:
I think the doctype should be before the root element. What do you think?

You could for instance split each line on `<` and:

1. if the first letter is `?`, then continue (probably the XML declaration)
2. if the string starts with `!DOCTYPE html>` => `return true;`
3. if the first letter is `!`, then continue (probably a comment)
4. if the split position is 1 or more => this is an element => `return false;`

I think that would work. Or maybe use a regex. Or something else :)

Reply:
Hi @josteinaj 

Sounds reasonable. A simple regex just checking if we have a `<` followed by a letter should be enough to shortcircuit the execution and improve execution, ensuring that the doctype is before the actual document. 

I also updated the test case and moved the doctype into the document.

Best regards
Daniel